config.org
#+title: Doom Emacs Configuration
#+subtitle: Taming the Tyranny of Text Transformation
#+author: Dadams
#+startup: content

* Table of Contents :TOC:
- [[#preamble][Preamble]]
- [[#the-simple-stuff][The simple stuff]]
  - [[#personal-info][Personal info]]
  - [[#default-settings][Default settings]]
  -  [[#doom-config][Doom Config]]
  - [[#fixing-bugs][Fixing bugs]]
  - [[#other-stuff][Other stuff]]
- [[#packages][Packages]]
  - [[#loading-instructions][Loading instructions]]
  - [[#convenience][Convenience]]
  - [[#tools][Tools]]
  - [[#visuals][Visuals]]
  - [[#file-types][File Types]]
- [[#applications][Applications]]
  - [[#mail][Mail]]
  - [[#rssnewsfeed-reader][RSS/Newsfeed Reader]]
- [[#languages][Languages]]
  - [[#org][Org]]
  - [[#latex][Latex]]
  - [[#python][Python]]
  - [[#graphviz][Graphviz]]
  - [[#regex][Regex]]
  - [[#hex][Hex]]
  - [[#markdown][Markdown]]
-   [[#the-rest][The rest]]
- [[#other-packages][Other packages]]

* Preamble

I don't really know what I am doing, but it seems fun. If you want an actual representation of something good check out [[https://tecosaur.github.io/emacs-config/config.html][Tecosaurs Config]].

A note to the unweary adventurer. At no point in this config should you put the line:
#+begin_src org :tangle no
#+ +begin_src emacs-lisp tangle: "yes"
#+end_src
Org assumes the =tangle: yes= by default for all source blocks but if you manually put it there this breaks that assumption
and no source blocks will be tangled


* The simple stuff

apparently this makes things run faster (I found it [[https://nullprogram.com/blog/2016/12/22/][here]])
#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+end_src


#+begin_src shell :exports none :comments no :tangle-mode (identity #o755)
#!/usr/bin/env bash
#+end_src

** Personal info

From ~GPG~ to ~git~ to email clients and snippets, the following seems kinda important to have around
#+begin_src emacs-lisp
(setq user-full-name "David Adams"
      user-mail-address "david@dadams2.com")
#+end_src


*** GPG
#+begin_src emacs-lisp
(setq auth-sources '("~/.authinfo.gpg")
      auth-source-cache-expiry nil) ; default is 7200 (2h)
#+end_src
** Default settings

*** Common settings
Following in

Tec's footsteps I somehow stumbled upon [[https://github.com/angrybacon/dotemacs/blob/master/dotemacs.org#use-better-defaults][angrybacon/dotemacs]] which gave some inspiration on better default settings

#+begin_src emacs-lisp
(setq-default
 delete-by-moving-to-trash t                    ; its rubbish bin not trash
 window-combination-resize t                    ; take all new space
 x-stretch-cursor t                             ; deal with glyphs
 ;; enable-local-variables t                    ; I want to be able to use dir-locals
 ;; global-auto-revert-mode t                   ; reload files on disk
 )


(setq
 undo-limit 80000000
 evil-want-fine-undo t                          ; make undo limit bigger
 auto-save-default t                            ; help with the b-s
 truncate-string-ellipsis "…"                   ; is nicer
 password-cache-expiry nil
 compilation-scroll-output t                    ; warnings annoy me
 )

(unless (string-match-p "^Power N/A" (battery)) ; On laptops...
  (display-battery-mode 1))                     ; it's nice to know how much power you have

;; (global-subword-mode 1)                         ; Iterate through CamelCase words
#+end_src

***  Frame sizing

Creating a new full-size window each time we want a new frame does not seem ideal. Especially considering window managers
make it easy to snap windows into the correct place

#+begin_src emacs-lisp
(add-to-list 'default-frame-alist '(height . 24))
(add-to-list 'default-frame-alist '(width . 80))
#+end_src

whilst this is not specifically needed for my setup on WM's, I do have to use normal desktop envs so this is useful.


***  Window Splitting

I always want new windows to be to the right

#+begin_src emacs-lisp
(setq evil-vsplit-window-right t
      evil-split-window-below t)
#+end_src
and when I do I would like to be promoted for what buffer to switch to


#+begin_src emacs-lisp
(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (consult-buffer))
#+end_src

being able to rotate window layouts is nice
#+begin_src elisp
(map! :map evil-window-map
      "SPC" #'rotate-layout)
#+end_src


*** Window switching
doom uses ace window for quick navigation between multiple windows.
By default numbers are used. I would rather use the home row
Simply don't' add src_elisp{+numbers} to src_elisp{ui: window-select}
#+begin_src emacs-lisp
;; (setq aw-keys '(?a ?s ?d ?f ?g ?h ?j ?k ?l))
;; (custom-set-faces!
;;   '(aw-leading-char-face
;;     :foreground "white" :background "red"
;;     :weight bold :height 2.5 :box (:line-width 5 :color "red")))
#+end_src
TODO: consider a better face for it
*** Embark for window switching
#+begin_src emacs-lisp
(after! embark
  (eval-when-compile
    (defmacro my/embark-ace-action (fn)
      `(defun ,(intern (concat "my/embark-ace-" (symbol-name fn))) ()
         (interactive)
         (with-demoted-errors "%s"
           (require 'ace-window)
           (let ((aw-dispatch-always t))
             (aw-switch-to-window (aw-select nil))
             (call-interactively (symbol-function ',fn)))))))

  (define-key embark-file-map     (kbd "o") (my/embark-ace-action find-file))
  (define-key embark-buffer-map   (kbd "o") (my/embark-ace-action switch-to-buffer))
  (define-key embark-bookmark-map (kbd "o") (my/embark-ace-action bookmark-jump))
  )
#+end_src
*** Macos specific
given how the commands usually change on macos this makes more sense to me
#+begin_src emacs-lisp
(setq mac-command-modifier 'control)
#+end_src


**  Doom Config

*** Modules
:PROPERTIES:
:header-args:emacs-lisp: :tangle no
:END:
You have probable seen a ~how to install~ section at the top of a bunch of emacs pages. Doom has it so that you remove a comment
and that's it. There are also extra features that can be added to some of the modules with a =+<somtehing>=

#+begin_src emacs-lisp :tangle "init.el" :noweb no-export :comments no

;;; init.el -*- lexical-binding: t; -*-

;; This file controls what Doom modules are enabled and what order they load
;; in. Remember to run 'doom sync' after modifying it!

;; NOTE Press 'SPC h d h' (or 'C-h d h' for non-vim users) to access Doom's
;;      documentation. There you'll find a "Module Index" link where you'll find
;;      a comprehensive list of Doom's modules and what flags they support.

;; NOTE Move your cursor over a module's name (or its flags) and press 'K' (or
;;      'C-c c k' for non-vim users) to view its documentation. This works on
;;      flags as well (those symbols that start with a plus).
;;
;;      Alternatively, press 'gd' (or 'C-c c d') on a module to browse its
;;      directory (for easy access to its source code).

(doom!  :input
        <<doom-input>>

        :completion
        <<doom-completion>>

        :ui
        <<doom-ui>>

        :editor
        <<doom-editor>>

        :term
        <<doom-term>>

        :checkers
        <<doom-checkers>>

        :tools
        <<doom-tools>>

        :os
        <<doom-os>>

        :lang
        <<doom-lang>>

        :email
        <<doom-email>>

        :app
        <<doom-app>>

        :config
        <<doom-config>>
        ) ;lol don't forget this closing brace like I did
#+end_src

**** Config's config

If you don't know what a literate configuration is, welcome! You are reading one.
Doom has some rather nice defaults for literate src_elisp{:config}
#+name: doom-config
#+begin_src emacs-lisp
literate
(default +bindings +smartparens)
#+end_src

**** Completion

#+name: doom-completion
#+begin_src emacs-lisp
(company            ; the ultimate code completion backend
 +childframe)       ; ... when the children are actually better
;;helm              ; the *other* search engine for love and life
;;ido               ; the other *other* search engine...
;;(ivy
;; +icons           ; a search engine for love and life
;;)
(vertico +icons)    ; search endine for the future
#+end_src

**** General UI

Some are doom specific whilst others are nice quality of life improvements

#+name: doom-ui
#+begin_src emacs-lisp
;;deft                 ; notational velocity for Emacs
doom                   ; what makes DOOM look the way it does
doom-dashboard         ; a nifty splash screen for Emacs
;;doom-quit              ; DOOM quit-message prompts when you quit Emacs
(emoji +unicode)       ; 🙂
hl-todo                ; highlight TODO/FIXME/NOTE/DEPRECATED/HACK/
;;hydra                ; quick documentation for related commands
;;indent-guides        ; highlighted indent columns
(ligatures +extra)     ; ligatures and symbols to make your code pretty again
minimap                ; show a map of the code on the side
modeline               ; snazzy, Atom-inspired modeline, plus API
nav-flash              ; blink cursor line after big
;;neotree              ; a project drawer, like NERDTree for vim
ophints                ; highlight the region an operation acts on
(popup +defaults +all) ; tame sudden yet inevitable temporary windows
tabs                   ; a tab bar for Emacs and prettyer tabs
treemacs               ; a project drawer, like neotree but cooler
unicode                ; extended unicode support for various languages
vc-gutter              ; vcs diff in the fringe
vi-tilde-fringe        ; fringe tildes to mark beyond EOB
window-select          ; visually switch windows
workspaces             ; tab emulation, persistence & separate workspaces
zen                    ; distraction-free coding or writing
#+end_src

**** Editor things

#+name: doom-editor
#+begin_src emacs-lisp
(evil +everywhere)     ; come to the dark side, we have cookies
file-templates         ; auto-snippets for empty files
fold                   ; (nigh) universal code folding
(format +onsave)       ; automated
;;god                  ; run Emacs commands without modifier keys
;;lispy                ; vim for lisp, for people who don't like vim
multiple-cursors       ; editing in many places at once
;;objed                ; text object editing for the innocent
;;parinfer             ; turn lisp into python, sort of
;;rotate-text          ; cycle region at point between text candidates
snippets               ; my elves. They type so I don't have to
;;word-wrap            ; soft wrapping with language-aware indent

:emacs
(dired +icons)         ; making dired pretty [functional]
electric               ; smarter, keyword-based electric-indent
(ibuffer +icons)       ; interactive buffer management
undo                   ; persistent, smarter undo for your inevitable mistakes
vc                     ; version-control and Emacs, sitting in a tree
#+end_src

**** Terminals

whilst vterm is the best terminal emulator inside emacs currently, it is the most painful thing to compile (with
                                                                                                            the exception of maybe pdftools). As such having alternatives can be useful

#+name: doom-term
#+begin_src emacs-lisp
eshell            ; the elisp shell that works everywhere
;;shell           ; simple shell REPL for Emacs
;;term            ; basic terminal emulator for Emacs
vterm             ; the best terminal emulation in Emacs
#+end_src

**** Checkers

not much to be said. spell-fu seems good but the interfaces are not as nice
grammar would be much better if it offered suggestions on how to improve things (well specifically writegood mode)
#+name: doom-checkers
#+begin_src emacs-lisp
syntax              ; tasing you for every semicolon you forget
(:if (executable-find "aspell") (spell +flyspell +aspell +everywhere)) ; tasing you for misspelling mispelling
grammar           ; tasing grammar mistake every you make
#+end_src


**** Other stuff

despite the name herein lies some of the things that make emacs worth using.

#+name: doom-tools
#+begin_src emacs-lisp
ansible
biblio        ; Writes a PhD for you (citation needed)
(debugger +lsp) ; FIXME stepping through code, to help you add bugs
;;direnv        ; be direct about your environment
docker        ; port everything to containers
editorconfig    ; let someone else argue about tabs vs spaces
;;ein           ; tame Jupyter notebooks with emacs
(eval +overlay) ; run code, run (also, repls)
;;gist          ; interacting with github gists
lookup          ; navigate your code and its documentation
(lsp +elgot)    ; M-x vscode
;;macos         ; MacOS-specific commands
(magit +forge)  ; a git porcelain for Emacs
make            ; run make tasks from Emacs
;;pass          ; password manager for nerds
pdf             ; pdf enhancements
;;prodigy       ; FIXME managing external services & code builders
rgb             ; creating color strings
;;taskrunner    ; taskrunner for all your projects
;;terraform     ; infrastructure as code
;;tmux          ; an API for interacting with tmux
tree-sitter     ; syntax and parsing, sitting in a tree...
upload          ; map local to remote projects via ssh/ftp
#+end_src
#+name: doom-os
#+begin_src emacs-lisp
(:if IS-MAC macos)  ; improve compatibility with macOS
tty                 ; improve the terminal Emacs experience
#+end_src


**** Languages

#+name: doom-lang
#+begin_src emacs-lisp
;;agda              ; types of types of types of types...
;;beancount         ; mind the GAAP
(cc +lsp)           ; C > C++ == 1
;;clojure           ; java with a lisp
;;common-lisp       ; if you've seen one lisp, you've seen them all
;;coq               ; proofs-as-programs
;;crystal           ; ruby at the speed of c
;;csharp            ; unity, .NET, and mono shenanigans
;;data              ; config/data formats
;;(dart +flutter)   ; paint ui and not much else
;;dhall
;;elixir            ; erlang done right
;;elm               ; care for a cup of TEA?
emacs-lisp          ; drown in parentheses
;;erlang            ; an elegant language for a more civilized age
;;ess               ; emacs speaks statistics
;;factor
;;faust             ; dsp, but you get to keep your soul
(fortran +lsp)      ; in FORTRAN, GOD is REAL (unless declared INTEGER)
;;fsharp            ; ML stands for Microsoft's Language
;;fstar             ; (dependent) types and (monadic) effects and Z3
;;gdscript          ; the language you waited for
;;(go +lsp)         ; the hipster dialect
;;(graphql +lsp)    ; Give queries a REST
(haskell)    ; a language that's lazier than I am
;;hy                ; readability of scheme w/ speed of python
;;idris             ; a language you can depend on
(json
 +lsp
 +tree-sitter)      ; At least it ain't XML
(java
 ;; +lsp
 +tree-sitter)      ; the poster child for carpal tunnel syndrome
(javascript +lsp)   ; all(hope(abandon(ye(who(enter(here))))))
;; (julia +lsp)     ; a better, faster MATLAB
;;kotlin            ; a better, slicker Java(Script)
(latex +lsp)        ; writing papers in Emacs has never been so fun
;;lean              ; for folks with too much to prove
;;ledger            ; be audit you can be
(lua +tree-sitter)  ; one-based indices? one-based indices
(markdown +grip)    ; writing docs for people to ignore
;;nim               ; python + lisp at the speed of c
nix               ; I hereby declare "nix geht mehr!"
;;ocaml             ; an objective camel
(org
 +pretty
 +pandoc
 +present
 +roam2
 +hugo
 +gnuplot
 +dragndrop
 +contacts
 +jupyter)          ; organize your plain life in plain text
;;php               ; perl's insecure younger brother
;;plantuml          ; diagrams for confusing people more
;;purescript        ; javascript, but functional
(python
 +lsp
 +pyright
 ;; +conda          ; micromamba is better :D
 +cython)           ; beautiful is better than ugly
;;qt                ; the 'cutest' gui framework ever
;;racket            ; a DSL for DSLs
;;raku              ; the artist formerly known as perl6
;;rest              ; Emacs as a REST client
;;rst               ; ReST in peace
;;(ruby +rails)     ; 1.step {|i| p "Ruby is #{i.even? ? 'love' : 'life'}"}
(rust +lsp)         ; Fe2O3.unwrap().unwrap().unwrap().unwrap()
;;scala             ; java, but good
;;(scheme +guile)   ; a fully conniving family of lisps
(sh +lsp)           ; she sells {ba,z,fi}sh shells on the C xor
;;sml
;;solidity          ; do you need a blockchain? No.
;;swift             ; who asked for emoji variables?
;;terra             ; Earth and Moon in alignment for performance.
web                 ; the tubes
(yaml +lsp)         ; JSON, but readable
;;zig               ; C, but simpler
#+end_src


**** Bring it all in

A constant work in progress. But trying to understand it does teach you a bunch.

#+name: doom-email
#+begin_src emacs-lisp
(:if (executable-find "mu") (mu4e +gmail))
;;notmuch
;;(wanderlust +gmail)
#+end_src

there are other things to consider as well
#+name: doom-app
#+begin_src emacs-lisp
;;calendar          ; A dated approach to timetabling
;;emms              ; Multimedia in Emacs is music to my ears
everywhere          ; *leave* Emacs!? You must be joking
irc                 ; how neckbeards socialize
(rss +org)          ; emacs as an RSS reader
;;twitter           ; twitter client https://twitter.com/vnought
#+end_src


**** Input

Whilst I could probably do japanese it would be a struggle so lets not
TODO: due to a bug in async tangle we don't include this
#+begin_src emacs-lisp :tangle no
;;bidi                       ; (tfel ot) thgir etirw uoy gnipleh
;;chinese
;;japanese
;;layout                     ; auie,ctsrnm is the superior home row
#+end_src

*** General Visual

**** Fonts

I am a big fan of 'Fira Code'. Font sizes do need some fiddling to get it all right.
#+begin_src emacs-lisp
(setq doom-font (font-spec :family "Fira Code" :size 12)
      doom-big-font (font-spec :family "Fira Code" :size 16)
      doom-variable-pitch-font (font-spec :family "Overpass" :size 16)
      doom-unicode-font (font-spec :family "JuliaMono")
      doom-serif-font (font-spec :family "IBM Plex Mono" :weight 'light)
      )
#+end_src

***** Missing Fonts

Warning about missing fonts is a bit of a problem especially if you are coming to a new system.
The following code (once again shamelessly ripped from Tec) makes it slightly more nice.
#+name: detect-missing-fonts
#+begin_src emacs-lisp :tangle no
(defvar required-fonts      '("Fira Code*" "Overpass" "JuliaMono" "IBM Plex Mono" "Merriweather" "Alegreya"))

(defvar available-fonts
  (delete-dups (or (font-family-list)
                   (split-string (shell-command-to-string "fc-list : family")
                                 "[,\n]"))))

(defvar missing-fonts
  (delq nil (mapcar
             (lambda (font)
               (unless (delq nil (mapcar (lambda (f)
                                           (string-match-p (format "^%s$" font) f))
                                         available-fonts))
                 font))
             required-fonts)))

(if missing-fonts
    (pp-to-string
     `(unless noninteractive
        (add-hook! 'doom-init-ui-hook
          (run-at-time nil nil
                       (lambda ()
                         (message "%s missing the following fonts: %s"
                                  (propertize "Warning!" 'face '(bold warning))
                                  (mapconcat (lambda (font)
                                               (propertize font 'face 'font-lock-variable-name-face))
                                             ',missing-fonts
                                             ", "))
                         (sleep-for 0.5))))))
  ";; No missing fonts detected")
#+end_src

#+begin_src emacs-lisp :noweb no-export
<<detect-missing-fonts()>>
#+end_src
***** Fixing icons

Some icons such as       (should see a windows icon, up down arrows a clock and a calendar not rain clouds a rain drop or stars) can appear to be incorrect depending on what font doom uses for unicode rendering.
This can make things difficult if you are working on something that renders them differently.
#+begin_src emacs-lisp
(setq doom-unicode-font (font-spec :family "Fira Code Nerd Font"))
#+end_src

**** Theme and modeline

I like the softer look of Nord and use it in a number of other places. There are some disadvantages like weird comments
but overall its nice.

#+begin_src emacs-lisp
(setq doom-theme 'doom-one)
#+end_src

I use linux on all my systems so I don't need to be told that =LF UTF-8= is my file encoding all the time
#+begin_src emacs-lisp
(defun doom-modeline-conditional-buffer-encoding ()
  "We expect the encoding to be LF UTF-8, so only show the modeline when this is not the case"
  (setq-local doom-modeline-buffer-encoding
              (unless (and (memq (plist-get (coding-system-plist buffer-file-coding-system) :category)
                                 '(coding-category-undecided coding-category-utf-8))
                           (not (memq (coding-system-eol-type buffer-file-coding-system) '(1 2))))
                t)))

(add-hook 'after-change-major-mode-hook #'doom-modeline-conditional-buffer-encoding)
#+end_src

there are also a few other nice things I like
#+begin_src elisp
(setq doom-modeline-minor-modes nil)
(setq doom-modeline-enable-word-count t) ; super nice in visual mode
(setq doom-modeline-persp-name t) ; I want to know where I am
#+end_src
**** Word wrapping
#+begin_src emacs-lisp
;; (setq +word-wrap-fill-style 'auto)
#+end_src

** Fixing bugs

in =mu= > 1.8 there is a problem with dooms default popup rules. Lets fix this
#+begin_src emacs-lisp
(set-popup-rule! "^\\*mu4e-\\(main\\|headers\\)\\*" :ignore t)
#+end_src




** Other stuff
*** Allow babel execution in CLI actions
In this config I sometimes generate code to include in my config.
This works nicely, but for it to work with =doom sync= et. al. I need to make sure
that Org doesn't try to confirm that I want to allow evaluation (I do!).

Thankfully Doom supports =$DOOMDIR/cli.el= file which is sourced every time a CLI
command is run, so we can just enable evaluation by setting
~org-confirm-babel-evaluate~ to ~nil~ there.
While we're at it, we should silence ~org-babel-execute-src-block~ to
avoid polluting the output.

#+begin_src emacs-lisp :tangle cli.el :comments no
;;; cli.el -*- lexical-binding: t; -*-
(setq org-confirm-babel-evaluate nil)

(defun doom-shut-up-a (orig-fn &rest args)
  (quiet! (apply orig-fn args)))

(advice-add 'org-babel-execute-src-block :around #'doom-shut-up-a)
#+end_src

*** Line numbers

I literally cannot function without relative lines numbers
#+begin_src emacs-lisp
(setq display-line-numbers-type 'visual)
#+end_src
I recently changed this to be visual as I want to be able to do vim style jumps with folds
see [[help:display-line-numbers-type]]

*** Bookmarks
I like having access to bookmarks so lets save them in sane places and have a keybinding

#+begin_src emacs-lisp
(setq bookmark-default-file "~/.doom.d/bookmarks") ; I like being able to store my bookmarks properly

(map! :leader
      (:prefix-map ("b" . "buffer")
                   "C-s" 'bookmark-save))
#+end_src
*** Buffer names

Why not take advantage of icons for the default buffer
#+begin_src emacs-lisp
(setq doom-fallback-buffer-name "► Doom"
      +doom-dashboard-name "► Doom")

#+end_src

*** Splash screen

re-occurring pain point as this is the first thing that you see when you open emacs. Finding both the right image at the right size and colour is difficult.
At the moment it is simple but I have bigger plans

#+begin_src emacs-lisp
(defvar fancy-splash-image-template
  (expand-file-name "misc/splash-images/blackhole-lines.svg" doom-private-dir)
  "Default template svg used for the splash image, with substitutions from ")
(defvar fancy-splash-image-nil
  (expand-file-name "misc/splash-images/transparent-pixel.png" doom-private-dir)
  "An image to use at minimum size, usually a transparent pixel")

(setq fancy-splash-sizes
      `((:height 700 :min-height 70 :padding (0 . 2) :template ,(expand-file-name "misc/splash-images/blackhole-lines-0.svg" doom-private-dir))
        (:height 500 :min-height 50 :padding (1 . 2) :template ,(expand-file-name "misc/splash-images/blackhole-lines-0.svg" doom-private-dir))
        (:height 440 :min-height 42 :padding (1 . 4) :template ,(expand-file-name "misc/splash-images/blackhole-lines-0.svg" doom-private-dir))
        (:height 400 :min-height 38 :padding (1 . 4) :template ,(expand-file-name "misc/splash-images/blackhole-lines-1.svg" doom-private-dir))
        (:height 350 :min-height 36 :padding (1 . 3) :template ,(expand-file-name "misc/splash-images/blackhole-lines-2.svg" doom-private-dir))
        (:height 300 :min-height 34 :padding (1 . 3) :template ,(expand-file-name "misc/splash-images/blackhole-lines-3.svg" doom-private-dir))
        (:height 250 :min-height 32 :padding (1 . 2) :template ,(expand-file-name "misc/splash-images/blackhole-lines-4.svg" doom-private-dir))
        (:height 200 :min-height 30 :padding (1 . 2) :template ,(expand-file-name "misc/splash-images/blackhole-lines-5.svg" doom-private-dir))
        (:height 100 :min-height 24 :padding (1 . 2) :template ,(expand-file-name "misc/splash-images/emacs-e-template.svg" doom-private-dir))
        (:height 0   :min-height 0  :padding (0 . 0) :file ,fancy-splash-image-nil)))
;; "list of plists with the following properties
;; :height the height of the image
;; :min-height minimum `frame-height' for image
;; :padding `+doom-dashboard-banner-padding' to apply
;; :template non-default template file
;; :file file to use instead of template")

(defvar fancy-splash-template-colours
  '(("$colour1" . keywords) ("$colour2" . type) ("$colour3" . warning) ("$colour4" . base8))
  "list of colour-replacement alists of the form (\"$placeholder\" . 'theme-colour) which applied the template")

(unless (file-exists-p (expand-file-name "theme-splashes" doom-cache-dir))
  (make-directory (expand-file-name "theme-splashes" doom-cache-dir) t))

(defun fancy-splash-filename (theme-name height)
  (expand-file-name (concat (file-name-as-directory "theme-splashes")
                            theme-name
                            "-" (number-to-string height) ".svg")
                    doom-cache-dir))

(defun fancy-splash-clear-cache ()
  "Delete all cached fancy splash images"
  (interactive)
  (delete-directory (expand-file-name "theme-splashes" doom-cache-dir) t)
  (message "Cache cleared!"))

(defun fancy-splash-generate-image (template height)
  "Read TEMPLATE and create an image if HEIGHT with colour substitutions as
   described by `fancy-splash-template-colours' for the current theme"
  (with-temp-buffer
    (insert-file-contents template)
    (re-search-forward "$height" nil t)
    (replace-match (number-to-string height) nil nil)
    (dolist (substitution fancy-splash-template-colours)
      (goto-char (point-min))
      (while (re-search-forward (car substitution) nil t)
        (replace-match (doom-color (cdr substitution)) nil nil)))
    (write-region nil nil
                  (fancy-splash-filename (symbol-name doom-theme) height) nil nil)))

(defun fancy-splash-generate-images ()
  "Perform `fancy-splash-generate-image' in bulk"
  (dolist (size fancy-splash-sizes)
    (unless (plist-get size :file)
      (fancy-splash-generate-image (or (plist-get size :file)
                                       (plist-get size :template)
                                       fancy-splash-image-template)
                                   (plist-get size :height)))))

(defun ensure-theme-splash-images-exist (&optional height)
  (unless (file-exists-p (fancy-splash-filename
                          (symbol-name doom-theme)
                          (or height
                              (plist-get (car fancy-splash-sizes) :height))))
    (fancy-splash-generate-images)))

(defun get-appropriate-splash ()
  (let ((height (frame-height)))
    (cl-some (lambda (size) (when (>= height (plist-get size :min-height)) size))
             fancy-splash-sizes)))

(setq fancy-splash-last-size nil)
(setq fancy-splash-last-theme nil)
(defun set-appropriate-splash (&rest _)
  (let ((appropriate-image (get-appropriate-splash)))

config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "David Adams"
      user-mail-address "dbadams@iinet.net.au")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-nord)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)


;;(setq lsp-pyright-venv-path "/node/bud82/envs/second")

(setq-default
 delete-by-moving-to-trash t ; use systems trashcan
 window-combination-resize t ; take all new space
 x-stretch-cursor t          ; deal with glyphs
 )


(setq
 undo-limit 2000000
 evil-want-fine-undo t          ; make undo limit bigger
 auto-save-default t            ; help with the b-s
 truncate-string-ellipsis "…"   ; is nicer
 compilation-scroll-output t    ; warnings annoy me
 )

(global-subword-mode 1)

(add-to-list 'default-frame-alist '(height . 24)) ; better frame sizing
(add-to-list 'default-frame-alist '(width . 80))  ;



(setq evil-vsplit-window-right t
      evil-split-window-below t)

(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (consult-buffer))



(defun doom-modeline-conditional-buffer-encoding ()
  "We expect the encoding to be LF UTF-8, so only show the modeline when this is not the case"
  (setq-local doom-modeline-buffer-encoding
              (unless (and (memq (plist-get (coding-system-plist buffer-file-coding-system) :category)
                                 '(coding-category-undecided coding-category-utf-8))
                           (not (memq (coding-system-eol-type buffer-file-coding-system) '(1 2))))
                t)))

(add-hook 'after-change-major-mode-hook #'doom-modeline-conditional-buffer-encoding)
(setq projectile-indexing-method 'native)

(setq doom-modeline-minor-modes nil)
(setq doom-modeline-enable-word-count t)

(after! flycheck
  (map! :leader
        (:prefix-map ("c" . "code")
         "x" flycheck-command-map)))

;; auto fullscreen
(if (eq initial-window-system 'x)                 ; if started by emacs command or desktop file
    (toggle-frame-maximized)
  (toggle-frame-fullscreen))

;; fancy splash

(defvar fancy-splash-image-template
  (expand-file-name "misc/splash-images/emacs-e-template.svg" doom-private-dir)
  "Default template svg used for the splash image, with substitutions from ")

(defvar fancy-splash-sizes
  `((:height 300 :min-height 50 :padding (0 . 2))
    (:height 250 :min-height 42 :padding (2 . 4))
    (:height 200 :min-height 35 :padding (3 . 3))
    (:height 150 :min-height 28 :padding (3 . 3))
    (:height 100 :min-height 20 :padding (2 . 2))
    (:height 75  :min-height 15 :padding (2 . 1))
    (:height 50  :min-height 10 :padding (1 . 0))
    (:height 1   :min-height 0  :padding (0 . 0)))
  "list of plists with the following properties
  :height the height of the image
  :min-height minimum `frame-height' for image
  :padding `+doom-dashboard-banner-padding' (top . bottom) to apply
  :template non-default template file
  :file file to use instead of template")

(defvar fancy-splash-template-colours
  '(("$colour1" . keywords) ("$colour2" . type) ("$colour3" . base5) ("$colour4" . base8))
  "list of colour-replacement alists of the form (\"$placeholder\" . 'theme-colour) which applied the template")

(unless (file-exists-p (expand-file-name "theme-splashes" doom-cache-dir))
  (make-directory (expand-file-name "theme-splashes" doom-cache-dir) t))

(defun fancy-splash-filename (theme-name height)
  (expand-file-name (concat (file-name-as-directory "theme-splashes")
                            theme-name
                            "-" (number-to-string height) ".svg")
                    doom-cache-dir))

(defun fancy-splash-clear-cache ()
  "Delete all cached fancy splash images"
  (interactive)
  (delete-directory (expand-file-name "theme-splashes" doom-cache-dir) t)
  (message "Cache cleared!"))

(defun fancy-splash-generate-image (template height)
  "Read TEMPLATE and create an image if HEIGHT with colour substitutions as
   described by `fancy-splash-template-colours' for the current theme"
  (with-temp-buffer
    (insert-file-contents template)
    (re-search-forward "$height" nil t)
    (replace-match (number-to-string height) nil nil)
    (dolist (substitution fancy-splash-template-colours)
      (goto-char (point-min))
      (while (re-search-forward (car substitution) nil t)
        (replace-match (doom-color (cdr substitution)) nil nil)))
    (write-region nil nil
                  (fancy-splash-filename (symbol-name doom-theme) height) nil nil)))

(defun fancy-splash-generate-images ()
  "Perform `fancy-splash-generate-image' in bulk"
  (dolist (size fancy-splash-sizes)
    (unless (plist-get size :file)
      (fancy-splash-generate-image (or (plist-get size :template)
                                       fancy-splash-image-template)
                                   (plist-get size :height)))))

(defun ensure-theme-splash-images-exist (&optional height)
  (unless (file-exists-p (fancy-splash-filename
                          (symbol-name doom-theme)
                          (or height
                              (plist-get (car fancy-splash-sizes) :height))))
    (fancy-splash-generate-images)))

(defun get-appropriate-splash ()
  (let ((height (frame-height)))
    (cl-some (lambda (size) (when (>= height (plist-get size :min-height)) size))
             fancy-splash-sizes)))

(setq fancy-splash-last-size nil)
(setq fancy-splash-last-theme nil)
(defun set-appropriate-splash (&rest _)
  (let ((appropriate-image (get-appropriate-splash)))
    (unless (and (equal appropriate-image fancy-splash-last-size)
                 (equal doom-theme fancy-splash-last-theme)))
    (unless (plist-get appropriate-image :file)
      (ensure-theme-splash-images-exist (plist-get appropriate-image :height)))
    (setq fancy-splash-image
          (or (plist-get appropriate-image :file)
              (fancy-splash-filename (symbol-name doom-theme) (plist-get appropriate-image :height))))
    (setq +doom-dashboard-banner-padding (plist-get appropriate-image :padding))
    (setq fancy-splash-last-size appropriate-image)
    (setq fancy-splash-last-theme doom-theme)
    (+doom-dashboard-reload)))

(add-hook 'window-size-change-functions #'set-appropriate-splash)
(add-hook 'doom-load-theme-hook #'set-appropriate-splash)


(setq +ligatures-extras-in-modes '(elisp-mode))
(setq +ligatures-in-modes '(not special-mode comint-mode eshell-mode term-mode vterm-mode Info-mode python-mode))
